modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/CacheFreeList.java

```diff
@@ -76,4 +86,188 @@ public CacheFreeList(
+         */
+        private Boolean updateTtl(
+            int cacheId,
+            long pageId,
+            long page,
+            long pageAddr,
+            DataPageIO io,
+            CacheDataRow row,
+            int itemId,
+            Boolean walPlc,
+            IoStatisticsHolder statHolder
+        ) throws IgniteCheckedException {
+            // Read payload to determine that the row is fragmented.
+            DataPagePayload data = io.readPayload(pageAddr, itemId, pageMem.realPageSize(grpId));
+            if (data.nextLink() == 0) {
+                // The data is not fragmented, so it is one page update.
+                io.updateExpirationTime(pageAddr, data.offset(), row);
+
+                if (needWalDeltaRecord(pageId, page, walPlc)) {
+                    // TODO IGNITE-5829
+                    // This record must contain only a reference to a logical WAL record with the actual data.
+                    int rowSize = data.payloadSize();
+
+                    byte[] payload = new byte[rowSize];
+
+                    // Reload the payload to get the actual data.
+                    data = io.readPayload(pageAddr, itemId, pageSize());
+
+                    assert data.payloadSize() == rowSize;
+
+                    PageUtils.getBytes(pageAddr, data.offset(), payload, 0, rowSize);
+
+                    wal.log(new DataPageUpdateRecord(
+                        cacheId,
+                        pageId,
+                        itemId,
+                        payload));
+                }
+
+                return Boolean.TRUE;
+            }
+
+            // The data is fragmented, so need to update the required fragment.
+            long nextLink = 0L;
+            boolean pageUpdated = false;
+            int scannedBytes = 0;
+            int updatedBytes = 0;
+
+            // We are under write lock here, so it is safe to read/write the first page.
+            boolean firstPage = true;
+
+            do {
+                final int cutItemId = firstPage ? itemId : PageIdUtils.itemId(nextLink);
+                final long curPageId = firstPage ? pageId : PageIdUtils.pageId(nextLink);
+
+                final long curPage = firstPage ? page : pageMem.acquirePage(grpId, curPageId, statHolder);
+
+                try {
+                    // It is safe to acquire a write lock on the next pages because we are under the write lock on the first page (head of
+                    // the entry). This means that reading of entry is not possible, since it requires holding a read lock on the first page
+                    // until all pages related to the entry are read. Moreover, following the current implementation of inserting a new
+                    // entry, the tail of the entry is always placed on empty pages so these pages cannot be used for other entries.
+                    long curPageAddr = firstPage ? pageAddr : pageMem.writeLock(grpId, curPageId, curPage);
```

Comment:
This still feels dangerous if expiry time itself is fragmented. What if reader already read first page and now it reeds the rest of the pages. Holding write lock on the head won't prevent it from potentially reading corrupted value. This should not be a problem in real life. I don't know how to avoid such issue